grabber/anchors.py

```diff
@@ -1,18 +1,35 @@
 from typing import Tuple
+import cv2
 import numpy as np
 from napari.layers.points import Points
-
+from scipy.spatial import distance
 
 class Anchors(Points):
     def add(self, coord: Tuple[int, int]) -> None:
         coord = round(coord[0]), round(coord[1])
         grabber = self.metadata['grabber']
-        y, x = grabber.costs.shape
-        if self.is_valid(coord, on_contour=True):
+        
+        if self.is_valid(coord, on_contour=False):
+            coord = self.closest_contour_point(coord)
             index = grabber.add(coord)
             self.data = np.insert(self.data, index, np.atleast_2d(coord), axis=0)
 
     def is_valid(self, coord: Tuple[int, int], on_contour: bool) -> bool:
         grabber = self.metadata['grabber']
+        if grabber is None:
+            return False
         y, x = grabber.costs.shape
         return 0 <= coord[0] < y and 0 <= coord[1] < x and grabber.contour[coord] == on_contour
+
+    def closest_contour_point(self, coord) -> Tuple[int, int]:
+        grabber = self.metadata['grabber']
+        contours, _ = cv2.findContours(grabber.contour.astype(np.uint8), cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_NONE)
+        contour = sorted(contours, key=cv2.contourArea)[-1]
+        contour = contour.reshape((contour.shape[0], contour.shape[2]))
+        contour[:,[0, 1]] = contour[:,[1, 0]]
```

Comment:
I think the coordinates of the contour could be obtained using `np.array(np.nonzero(grabber.contour)).T` instead of selecting the contours with opencv since it is returning multiple contours.

Reply:
You're right, that approach is better